# Lab 3 - Rong Xu(Cherie)

In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os
import statsmodels.api as sm
import statsmodels.formula.api as smf

In [6]:
df = pd.read_csv("GSS.2006.csv")
df.head(5)

vpsu  vstrat  adults  ballot  dateintv  famgen  form  formwt  gender1  \
0     1    1957       1       3       316       2     1       1        2   
1     1    1957       2       2       630       1     2       1        2   
2     1    1957       2       2       314       2     1       1        2   
3     1    1957       1       1       313       1     2       1        2   
4     1    1957       3       1       322       2     2       1        2   

   hompop  ...  away7  gender14  old14  relate14  relhh14  relhhd14  relsp14  \
0       3  ...    NaN       NaN    NaN       NaN      NaN       NaN      NaN   
1       2  ...    NaN       NaN    NaN       NaN      NaN       NaN      NaN   
2       2  ...    NaN       NaN    NaN       NaN      NaN       NaN      NaN   
3       1  ...    NaN       NaN    NaN       NaN      NaN       NaN      NaN   
4       3  ...    NaN       NaN    NaN       NaN      NaN       NaN      NaN   

   where12  where6  where7  
0      NaN     NaN     NaN  
1      NaN     NaN     NaN  
2      NaN     NaN     NaN  
3      NaN     NaN     NaN  
4      NaN     NaN     NaN  

[5 rows x 1261 columns]

# 1. Run a simple bivariate regression, and interpret your results. (Did the results fit your expectations? Why? Why not?)

In [7]:
my_tab = pd.crosstab(index=df["stayorg3"],
                  columns="count")
def compute_percentage(x):
    pct = float(x/my_tab['count'].sum()) * 100
    return round(pct, 2)
my_tab['percentage'] = my_tab.apply(compute_percentage, axis=1)
my_tab

col_0     count  percentage
stayorg3                   
1.0         128       12.66
2.0         201       19.88
3.0         181       17.90
4.0         314       31.06
5.0         187       18.50

### I would turn down another job that offered quite a bit more pay in order to stay with this organization (STAYORG3)    1) strongly agree to 5) strongly disagree
### from the tab, around 18% are standing in netural that they are not sure if they will turn down the higher paid in order to stay; 49.56% disagree with the statement that they wont reject ; and nearly 33% agree with the statement that they will turn down. 

In [8]:
my_tab = pd.crosstab(index=df["bossemps"],
                  columns="count")
def compute_percentage(x):
    pct = float(x/my_tab['count'].sum()) * 100
    return round(pct, 2)
my_tab['percentage'] = my_tab.apply(compute_percentage, axis=1)
my_tab

col_0     count  percentage
bossemps                   
1.0         334       33.04
2.0         360       35.61
3.0         232       22.95
4.0          61        6.03
5.0          24        2.37

### In general, how would you describe relations in your work place between management and employees? (BOSSEMPS)    1) very good to 5) very bad 
### from the tab, around 22.95% are standing in netural; 68.65% think there is a good relation in their work place between management and employees; 8.4% think there is a bad relation.

In [9]:
lm = smf.ols(formula = 'stayorg3~bossemps', data = df).fit()
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               stayorg3   R-squared:                       0.143
Model:                            OLS   Adj. R-squared:                  0.142
Method:                 Least Squares   F-statistic:                     167.1
Date:                Fri, 03 Nov 2023   Prob (F-statistic):           1.85e-35
Time:                        17:10:18   Log-Likelihood:                -1618.0
No. Observations:                1006   AIC:                             3240.
Df Residuals:                    1004   BIC:                             3250.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.2035      0.088     25.018      0.000       2.031       2.376
bossemps       0.4909      0.038     12.925      0.000       0.416       0.565
==============================================================================
Omnibus:                       87.587   Durbin-Watson:                   1.969
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               35.319
Skew:                          -0.232   Prob(JB):                     2.14e-08
Kurtosis:                       2.208   Cond. No.                         6.19
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### For this question, I will buidl a bivariate regression for variables BOSSEMPS and STAYORG3. Based on the question description, I would summarize that BOSSEMPS is for workplace relationship and STAYORG3 is for staying or not with higher paid offered. In the bivariate regression, the independent variable is BOSSEMPS and dependent variable is STAYORG3. 
### I want to see how the relations in workplace between management and employees will affect the willingness of employee to stay that they turn down the higher piad job, by comparing these two variables, it can tell me the correlation between the workplace relationship and a willingness to stay. 

### My expectation is there is a positive relationship that people who think there is a good relation in their workplace will lead them to make choice of rejecting the higher paid from aother job in order to stay with the organzition. In order to verify it, I runned a bivariable regression model. 

### From the output, the R-square is 0.143, which means 14.3% of the variation in stay organization is explained by the workplace relations. Looking at coeffieient(bossemps), With 1 unit of independent variable bossemps (work relation) increase that meaning people are reducing their satisfaction on workplace relations, there is 0.4909 unit increasing in dependent variable stayorg3 (stay organization) which means people are more unlikely not turning down another higher paid job offered, with a statistically significant as pvalue is lower than 0. Looking at t score, t=12.925, from the normal distribution, we could know it has a significantly low possibily to have had happen by chance. It's reasonably unlikely that I would get a coefficient this large (in absolute terms) with a small standard error by chance. Therefore, there have a difference between bossemps and stayorg3 that bossemps could positively affect the stayorg3.

# 2. Add an additional variable that might mediate or partly "explain" the initial association from that simple regression above -- and explain your results. Did it work out? Yes? No?



In [10]:
my_tab = pd.crosstab(index=df["jobsat"],
                  columns="count")
def compute_percentage(x):
    pct = float(x/my_tab['count'].sum()) * 100
    return round(pct, 2)
my_tab['percentage'] = my_tab.apply(compute_percentage, axis=1)
my_tab

col_0   count  percentage
jobsat                   
1.0       187       18.41
2.0       381       37.50
3.0       304       29.92
4.0        69        6.79
5.0        31        3.05
6.0        31        3.05
7.0        13        1.28

### 1) completely satisfied to 7) completely dissatisfied, 4) neither satisfied nor dissatisfied
### from the tab, around 6.79% are standing in the middle that they are neither satisified nor dissatified their job; around 7.38% are not satisfied their job; most people are satisfied their job in general. 

In [11]:
lm2 = smf.ols(formula = 'stayorg3 ~ bossemps + jobsat', data = df).fit()
lm2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               stayorg3   R-squared:                       0.233
Model:                            OLS   Adj. R-squared:                  0.231
Method:                 Least Squares   F-statistic:                     152.0
Date:                Fri, 03 Nov 2023   Prob (F-statistic):           2.21e-58
Time:                        17:10:18   Log-Likelihood:                -1562.3
No. Observations:                1006   AIC:                             3131.
Df Residuals:                    1003   BIC:                             3145.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.7417      0.094     18.604      0.000       1.558       1.925
bossemps       0.2772      0.041      6.762      0.000       0.197       0.358
jobsat         0.3591      0.033     10.841      0.000       0.294       0.424
==============================================================================
Omnibus:                       47.227   Durbin-Watson:                   1.972
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               22.518
Skew:                          -0.153   Prob(JB):                     1.29e-05
Kurtosis:                       2.334   Cond. No.                         9.89
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### related question: How satisfied are you in your job? (JOBSAT) (job satisfaction)
### Adding an additional variable: jobsat;  workplace relationship lead to job satisfaction which lead to turn down for another job.
### from the output, the R-square is 0.233, which means 23.3% of the variation in dependent variable of staying organization is explained by the independent variables job satisfaction and workplace relationship. R-square ratio is greater than prior model that showing jobsat this additional variable is helpful to improve the model. In order to further proven my statement, let's also look at the other rate. 
### looking at the coeffeient, if constant the bossemps, there is a positive relationship between jobsat and stayorg3 that there is 0.3591 unit of stayorg3 increase for each one additional unit increase on jobsat, with a statistically significant as pvalue is lower than 0, which means people more tend to not turn down the higher paid offer when they tend to say they are not satisfied their job. looking at the t score, t=10.841, from the normal distribution, we could know it has a significantly low possibily to have had happen by chance. It's reasonably unlikely that I would get a coefficient this large (in absolute terms) with a small standard error (0.033) by chance. Therefore, there have a difference between jobsat and stayorg3 that bossemps could positively affect the stayorg3.
### If constant the jobsat, there is a positive relationship between satyorg3 and boosemps that there is 0.2772 unit of stayorg3 increase for each one addtional unit increase on bossempt, with a statistically significant as pvalue is lower than 0, which means people more tend to not turn down the higher paid offer when they tend to disagree their workplace had a good relationship between management and employees. Compared with 0.4909 in the bivariate regression, this coeffieent of bossemps is reduced to 0.2772, which could proven that addtional variable did improve the model. Looking at the t score, t=6.762, from the normal distribution, we could know it has a significantly low possibily to have had happen by chance. It's reasonably unlikely that I would get a coefficient this large (in absolute terms) with a small standard(0.041) error by chance. Therefore, there have a difference between bossemps and stayorg3 that bossemps could positively affect the stayorg3.
### for simple regression coeff(bossempt)1=0.4909, from the multiple regression  coeff(bossempt)2=0.2772, it reduced by 43.53% ((0.4909-0.2772)/0.4909), while its not reduce that large and very much, we could still say this model mediate and partly explain the initial association from that simple regression by following chain mechains, and there could have other important main mediating factors. 


# 3. Run another multiple regression. Tell me how you expect your dependent variable to be affected by the independent variables. Interpret your results.

related questions:
Respondent's education (EDUC)
Respondent's mother's education (MAEDUC)
Respondent's father's education (PAEDUC)
What is your race? (RACE)

In [18]:
## subset the race that be ready for question 4
dd = df.dropna(subset = ["race"])

In [19]:
lm3 = smf.ols(formula = 'educ ~ maeduc + paeduc', data = dd).fit()
lm3.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   educ   R-squared:                       0.306
Model:                            OLS   Adj. R-squared:                  0.306
Method:                 Least Squares   F-statistic:                     472.0
Date:                Fri, 03 Nov 2023   Prob (F-statistic):          1.75e-170
Time:                        18:47:17   Log-Likelihood:                -5115.7
No. Observations:                2139   AIC:                         1.024e+04
Df Residuals:                    2136   BIC:                         1.025e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      8.7412      0.176     49.696      0.000       8.396       9.086
maeduc         0.2131      0.020     10.714      0.000       0.174       0.252
paeduc         0.2311      0.018     13.072      0.000       0.196       0.266
==============================================================================
Omnibus:                       75.532   Durbin-Watson:                   1.823
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              191.713
Skew:                          -0.128   Prob(JB):                     2.34e-42
Kurtosis:                       4.444   Cond. No.                         52.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### for this multiple regression model, I wanted to study how the parents' education will affect their child's education. For my expectation, I assume that the null hypothesis is the parents' education will not affect their child's education. In order to verify it, let's look the output. 
### In the output, R-square =0.306, which means 30.6% of variation of child' education is explained by their  parent education from both side. Constant the father education, there is positive affect between educ and maeduc, on average, with additional one unit of mother education increase, their child education will increase 0.2131 unit, with a statistically significant.  t-stats = 10.714, which means there is low posibility that it have had happen by chance with this small standard error. Therefore, there has difference between educ and maeduc. 
### Constant the mother education, there is positive affect between educ and paeduc. On average, with additional one unit of father education increase, their child education will increase 0.2311 unit, with a statistically significant.  t-stats = 13.072, which means there is low posibility that it have had happen by chance with this small standard error. Therefore, there has difference between educ and paeduc. In conclusion, we could reject the null hypothesis that the parents education has positive affect on their child's education.

# 4. Now add another independent variable to that model in Question 3, specifically a set of dummy variables. Tell me why you added that new set of variables and what effect you expected them to have. Did they have an effect? Interpret that new model. 

In [21]:
## all respondent answers that under race level 
ddsub = pd.concat([dd.educ, dd.maeduc, dd.paeduc, pd.get_dummies(dd.race.astype(int), prefix="race")], axis=1)
ddsub.columns

Index(['educ', 'maeduc', 'paeduc', 'race_1', 'race_2', 'race_3'], dtype='object')

In [22]:
lm5 = smf.ols(formula = 'educ ~ maeduc + paeduc + race_1 + race_2 + race_3', data = ddsub).fit()
lm5.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   educ   R-squared:                       0.310
Model:                            OLS   Adj. R-squared:                  0.309
Method:                 Least Squares   F-statistic:                     239.8
Date:                Fri, 03 Nov 2023   Prob (F-statistic):          3.14e-170
Time:                        18:49:50   Log-Likelihood:                -5110.1
No. Observations:                2139   AIC:                         1.023e+04
Df Residuals:                    2134   BIC:                         1.026e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      6.5492      0.135     48.549      0.000       6.285       6.814
maeduc         0.1992      0.020      9.755      0.000       0.159       0.239
paeduc         0.2288      0.018     12.935      0.000       0.194       0.263
race_1         2.4831      0.101     24.541      0.000       2.285       2.682
race_2         2.1336      0.150     14.193      0.000       1.839       2.428
race_3         1.9325      0.128     15.085      0.000       1.681       2.184
==============================================================================
Omnibus:                       70.317   Durbin-Watson:                   1.827
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              175.538
Skew:                          -0.107   Prob(JB):                     7.63e-39
Kurtosis:                       4.387   Cond. No.                     1.01e+16
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 6.07e-27. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

### Now this model added in another variable race referring to race that people consider they are. Set it as dummy variable that there are 3 categories in total of one set. 1 refers to white, 2 refers to black or african american, and 3 are other race group. With the new set of dummy variables add in, R-square = 0.310 that rasie by 0.004, which helped to better study the question, also it means 31% of variation in child education could explained by all independent variables. 
### Constant other than father education variable, there is positive affect between educ and paeduc. On average, with additional one unit of father education increase, their child education will increase 0.2288 unit, with a statistically significant.  t-stats = 12.936, which means there is low posibility that it have had happen by chance with this small standard error. Therefore, there has difference between educ and paeduc. 
### Constant other than mother education variable, , there is positive affect between educ and maeduc, on average, with additional one unit of mother education increase, their child education will increase 0.1992 unit, with a statistically significant.  t-stats = 9.755, which means there is low posibility that it have had happen by chance with this small standard error. Therefore, there has difference between educ and maeduc. 
### while with the set of dummy variables addin, the both coeffienct of paeduc and maeduc are reduced so that this helped to better examinate the question. 
### I think prehaps people who with difference race could behavor different on their education in terms of their parents education. In general, as the race categories go higher, the lower the score on education with coeff from 2.4831 to 1.9325, although they all positive number. Compared to the reference category, individuals in the race category represented by race_1 white have, on average, an increase in white child education level by 2.4831 units while holding other variables constant with a statistically significant, while for race_2 black group, an increase in black child education level by 2.1336 units while holding other variables constant with a statistically significant. Therefore, we could say with increased family's parents education level, the white child's education level will rise better than black child's education on average with the absolute term. 

In [24]:
# the effect of adding dummy variable
sm.stats.anova_lm(lm3, lm5)

df_resid           ssr  df_diff    ss_diff        F    Pr(>F)
0    2136.0  14964.919996      0.0        NaN      NaN       NaN
1    2134.0  14886.794458      2.0  78.125538  5.59959  0.003754

### pvalue is close to 0 with a statistically significant, we could conclude that race is a significant predictor of did people's parents' education will impact their education level. 